### PR TITLE
MGDOBR-222: fix manager start with disabled rhoas (pt.2)

### PR DESCRIPTION
**Pull request by kostola**
[MGDOBR-222](https://issues.redhat.com/browse/MGDOBR-222) 

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### Description

This is the cleaner and final solution to the bug.

#### Major changes
*  `RhoasClient` and dependent beans are now produced using `Producer` methods, that take into account the value of the feature flag.
* Properties are read programmatically by the `Producer` methods only when the feature flag is enabled and passed to the beans via constructor, instead of using `@ConfigProperty`. This permits empty properties without errors if the feature flag is disabled.
* Added tests for all the possible scenarios: *"feature flag disabled"*, *"feature flag enabled"*, *"feature flag enabled with missing properties"* and *"default feature flag value (no explicit property set)"* with related [test profiles](https://quarkus.io/blog/quarkus-test-profiles/) that allow to spot unwanted configuration errors.

### CODE REVIEW

**Review by robpblake — COMMENTED** [review]
@kostola I wasn't in the call where you guys decided on this, so perhaps I'm missing something. I understand what you're trying to do here, but from my perspective I don't know if we need to go this far with it. 

If I were to summarise what I think is the problem:

1. We have environments where we want to run with dynamic topic provisioning enabled
2. We have environments where we don't want to run with dynamic topic provisioning enabled

So for example we do want to run with it enabled in our demo cluster and we will also want to enable it very soon in our CI pipelines so that we can test this functionality. We don't (yet) want to run with it enabled in our local dev because we don't have a good way of mocking out the Managed Kafka service.

However, we should eliminate that difference as quickly as we can so that we're always running the same code path locally as we do in production (or as closely as we possibly can). So for me, until we solve the "how do we mock Managed Kafka" problem,  the simplest solution here would be:

1. In environments where we want this functionality turned on, configuration values are required and must be provided by `ENV` variables as we do for everything else
2. In environments where we want this functionality turned off, use the simplest way of providing values for the configuration until we turn the functionality on. 

For scenario 2, the simplest thing to me seems to be to provide profile specific overlays e.g `%dev.event-bridge.rhoas.mgmt-api.host=invalid` 

Were there other discussions in the call that make my assumptions here invalid?

File: rhoas-client/src/main/java/com/redhat/service/bridge/rhoas/RhoasProducer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @ConfigProperty(name = ENABLED_FLAG, defaultValue = ENABLED_FLAG_DEFAULT_VALUE)
+    boolean rhoasEnabled;
+
+    @Produces
+    @ApplicationScoped
+    public RhoasClient produceRhoasClient(KafkasMgmtV1Client mgmtClient, KafkaInstanceAdminClient instanceClient) {
+        if (!rhoasEnabled) {
+            throw new RuntimeException(RHOAS_DISABLED_ERROR_MESSAGE);
+        }
```
Comment (r00ta) [inline rhoas-client/src/main/java/com/redhat/service/bridge/rhoas/RhoasProducer.java:32]: Is `rhoasEnabled=false` an exception?

**Comment by kostola** [comment]
@robpblake @r00ta @danielezonca @lucamolteni just to be clear: I introduced this "feature flag" to enable RHOAS via opt-in in #196 and I explained why [in this comment](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/196#discussion_r778940143).

We later realized it wasn't working as expected, because even if the flag was `false`, the absence of some RHOAS properties was crashing the manager (the exact behavior I was trying to avoid with the introduction of the flag).

The purpose of this PR is to fix the problem with a cleaner design and no need for setting dummy property values (like the quick fix in #213). Imho the effort with test profiles will be worth also in the future for other scenarios where we need to test different configurations.

In any case, this flag is **temporary**. It will be removed when the RHOAS integration will become mandatory (now it is not because the created topic is ignored) and we figure out how to test locally. Nothing more than this.

**Comment by robpblake** [comment]
> @robpblake @r00ta @danielezonca @lucamolteni just to be clear: I introduced this "feature flag" to enable RHOAS via opt-in in #196 and I explained why [in this comment](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/196#discussion_r778940143).
> 
> We later realized it wasn't working as expected, because even if the flag was `false`, the absence of some RHOAS properties was crashing the manager (the exact behavior I was trying to avoid with the introduction of the flag).
> 
> The purpose of this PR is to fix the problem with a cleaner design and no need for setting dummy property values (like the quick fix in #213). Imho the effort with test profiles will be worth also in the future for other scenarios where we need to test different configurations.
> 
> In any case, this flag is **temporary**. It will be removed when the RHOAS integration will become mandatory (now it is not because the created topic is ignored) and we figure out how to test locally. Nothing more than this.

@kostola OK, thanks for the explanation.

Can I ask you to make sure that the JIRA task we have to remove the feature flag has a complete list of everything here that needs to be cleaned up. Just in-case you're not the person that works on that task.

There are a couple of other comments on the PR that I think would help to clarify next steps on this work.

cheers,

Rob

File: manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java [hunk 2]
```diff
@@ -183,21 +183,22 @@ public BridgeResponse toResponse(Bridge bridge) {
...
-            } catch (CompletionException e) {
-                String msg = "Failed creating topic and service account for bridge " + bridgeId;
-                LOGGER.warn("[manager] " + msg, e);
-                throw new InternalPlatformException(msg, e);
-            } catch (TimeoutException e) {
-                String msg = "Timeout reached while creating topic and service account for bridge " + bridgeId;
-                LOGGER.warn("[manager] " + msg, e);
-                throw new InternalPlatformException(msg, e);
-            }
+        if (!rhoasEnabled) {
```
[the remark this answers is not in the document]
Comment (kostola) [reply manager/src/main/java/com/redhat/service/bridge/manager/BridgesServiceImpl.java:185]: Explained the purpose of the flag [here](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/218#issuecomment-1012157712)

**Comment by kostola** [comment]
> JIRA task we have to remove the feature flag

The cleanup task is [MGDOBR-225](https://issues.redhat.com/browse/MGDOBR-225)

**Review by lucamolteni — APPROVED** [review]

**Review by robpblake — APPROVED** [review]

**Pull request merged**